### ISSUE TITLE
Add support for excluding violations/files on negative regular expression match

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#3523]
Taken from #220.
User wants an option to specifically specify the list of files Checkstyle will check for violations by a regular expression. He wants to specify a regular expression that will ignore files if they don't match.

**Comment by rnveach** [comment]
I stopped at giving examples because I am unsure if the original request wants a special configuration for files with different names, or one configuration and just have certain checks suppressed.

There are multiple ways we could incorporate this if we wanted to.
* We could add negative regular expressions to suppressions with a new attribute.
* `BeforeExecutionExclusionFileFilter` could get a new option like `RegexpOnFilename` to do negative file matching and exclusions.
* CLI could be updated for a new command to do negative exclusion.
* We could add a file path expression to be customized for each individual fileset/check.

I am currently leaning towards the suppression route as it doesn't require a new configuration for a special type of file.
Example: `<suppress checks="SomeCheck" filesNotMatch="Remote\.java$"/>`

**Comment by pilhuhn** [comment]
This double negation suppress for notMatch is complicated to get around in the head.
I think having this as positive wording `<include checks="SomeCheck" filesMatch="Remote\.java$"/>`would be a lot better

**Comment by rnveach** [comment]
@pilhuhn the current `files` attribute is already a matcher.
`<include files="Remote\.java$"/>` = Suppress all files that match "*Remote.java"

You want a regular expression that does the inverse of `files`, hence why I went with the name `filesNotMatch`.

**Comment by romani** [comment]
I removed label "approved" as requirements and implementation is not clear.

**Comment by pilhuhn** [comment]
@romani Yes. Your above was `<suppress`.. and not `<include` which would be a double negation

**Comment by rnveach** [comment]
That was my bad copying the `<include` for showing `files` attribute in my example. I meant `<suppress`. I didn't even notice you had `include` there.

How would the `checks` attribute work with `<include`? It would still be matching while `file` is the negation match?

**Comment by MEZk** [comment]
@romani @rnveach 
Not sure whether we need this fix as we are goind to reimplement suppresion model and maybe make it similar to PMD's. The fix might be reasonable as a temporal solution.

**Comment by strkkk** [comment]
I can suggest creating a new filter, which will have name like BeforeExecutionIncludedFileFilter (or similar) and there are going to be regex for files names to include.

Some negative regexps/properties addition to existing filter will be confusing, suppression approach have a problem - if we need to inspect a small subset of files in big codebase, checkstyle will do a lot of unnecessary work to produce and suppress violations.

**Comment by romani** [comment]
can we use negative lookahead ? https://stackoverflow.com/a/1240365
it is not pleasant syntax to read but it should work even now.
`filesNotMatch` a ways more human friendly approach.

> and there are going to be regex for files names to include.

in this case it is better to configure CLI or plugin to target only certain files.

**Comment by strkkk** [comment]
> `filesNotMatch` a ways more human friendly approach.

Name of the filter is BeforeExecution**Exclusion**FileFilter, so it is confusing to have there some regex for files to **include** in analysis

**Comment by romani** [comment]
@rnveach , lets close this issue and start discussion on exact Filter, looks like we are talking about different issues.

**Comment by rnveach** [comment]
You can also discuss issue here and create new issue on finalized direction.

**Comment by romani** [comment]
original issue is for file filtering by negation.
it could be done as:
```
$ cat Test.java 
public class Test {
  String value = "\u03bc\b"; // ok
}

$ cat MyRemote.java 
public class Test {
  String value = "\u03bc\b"; // ok
}

$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="BeforeExecutionExclusionFileFilter">
      <property name="fileNamePattern" value="^((?!Remote\.java).)*$"/>
    </module>

    <module name="FileLength">
      <property name="max" value="1"/>
    </module>
</module>

$ java -jar checkstyle-8.38-all.jar -c config.xml Test.java MyRemote.java 
Starting audit...
[ERROR] /var/tmp/MyRemote.java:1: File length is 3 lines (max allowed is 1). [FileLength]
Audit done.
Checkstyle ends with 1 errors.
```

**Comment by romani** [comment]
@rnveach , the same approach could be applied to all regexp that we have, so user can make it happen even now, it will not very user friendly, but doable and still not clear why they might need need such negation so frequently.
if they need - they have ability already.

I think we can close this issue till more reasons are shared.

**Comment by rnveach** [comment]
if negative lookahead is how we want to proceed, then examples is going to be the core of explaining this to the users.

> "^((?!Remote\.java).)*$"
> [ERROR] /var/tmp/MyRemote.java:1: File length is 3 lines (max allowed is 1).

This example doesn't make much sense to me. I assume you are trying to configure suppression to suppress all files but exactly `Remote.java`?

None of your files are named `Remote.java` only (since you are specifying `^` and `$`) and the `.` and `*` are throwing me off on what this is trying to accomplish and how it works.

If it is looking to suppress all files but exactly `*Remote.java`, then how would I configure it for exactly `Remote.java` and not `MyRemote.java`?

Another thing I can see people wanting is, how would I configure it to suppress everything else except `Remote` and `Client`?

**Comment by romani** [comment]
@rnveach,
```
$ wc -l Test.java MyRemote.java MyClient.java Client.java Remote.java
  3 Test.java
  3 MyRemote.java
  3 MyClient.java
  3 Client.java
  3 Remote.java
 15 total

$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="BeforeExecutionExclusionFileFilter">
      <property name="fileNamePattern" value="^(?!.*(Remote\.java|Client\.java|[\\/]Remote\.java|[\\/]Client\.java)).*$"/>
    </module>

    <module name="FileLength">
      <property name="max" value="1"/>
    </module>
</module>

$ java -jar checkstyle-8.38-all.jar -c config.xml Test.java MyRemote.java MyClient.java Client.java Remote.java 
Starting audit...
[ERROR] /var/tmp/MyRemote.java:1: File length is 3 lines (max allowed is 1). [FileLength]
[ERROR] /var/tmp/MyClient.java:1: File length is 3 lines (max allowed is 1). [FileLength]
[ERROR] /var/tmp/Client.java:1: File length is 3 lines (max allowed is 1). [FileLength]
[ERROR] /var/tmp/Remote.java:1: File length is 3 lines (max allowed is 1). [FileLength]
Audit done.
Checkstyle ends with 4 errors.
```

is it what you wanted ?
